compiler/src/Language/Bond/Codegen/Cs/Grpc_cs.hs

```diff
@@ -0,0 +1,147 @@
+-- Copyright (c) Microsoft. All rights reserved.
+-- Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+{-# LANGUAGE QuasiQuotes, OverloadedStrings, RecordWildCards #-}
+
+module Language.Bond.Codegen.Cs.Grpc_cs (
+  grpc_cs)
+where
+
+import Data.Maybe
+import Data.Monoid
+import Prelude
+import qualified Data.Text.Lazy as L
+import Text.Shakespeare.Text
+import Language.Bond.Syntax.Types
+import Language.Bond.Util
+import Language.Bond.Codegen.Util
+import Language.Bond.Codegen.TypeMapping
+import qualified Language.Bond.Codegen.Cs.Util as CS
+
+-- | Codegen template for generating code containing declarations of
+-- of services including proxies and services files for gRPC integration.
+
+grpc_cs :: MappingContext -> String -> [Import] -> [Declaration] -> (String, L.Text)
+grpc_cs cs _ _ declarations = ("_grpc.cs", [lt|
+#{CS.disableCscWarnings}
+#{CS.disableReSharperWarnings}
+
+namespace #{csNamespace}
+{
+    #{doubleLineSep 1 grpc declarations}
+} // #{csNamespace}
+|])
+  where
+    csNamespace = sepBy "." toText $ getNamespace cs
+    idl = MappingContext idlTypeMapping [] [] []
+
+    grpc s@Service{..} = [lt|#{CS.typeAttributes cs s}public static class #{declName}#{generics} #{genericsWhere}
+    {
+        static readonly string ServiceName = "#{getDeclTypeName idl s}";
+
+        #{doubleLineSep 2 methodDeclaration serviceMethods}
+
+        public abstract class #{declName}Base
+        {
+            #{doubleLineSep 3 serverBaseMethods serviceMethods}
+        }
+
+        public class #{declName}Client : global::Grpc.Core.ClientBase<#{declName}Client>
+        {
+            public #{declName}Client(global::Grpc.Core.Channel channel) : base(channel)
+            {
+            }
+
+            protected #{declName}Client() : base()
+            {
+            }
+
+            protected #{declName}Client(global::Grpc.Core.ClientBase.ClientBaseConfiguration configuration) : base(configuration)
+            {
+            }
+
+            #{doubleLineSep 3 clientMethods serviceMethods}
+
+            protected override #{declName}Client NewInstance(global::Grpc.Core.ClientBase.ClientBaseConfiguration configuration)
+            {
+                return new #{declName}Client(configuration);
+            }
+        }
+
+        public static global::Grpc.Core.ServerServiceDefinition BindService(#{declName}Base serviceImpl)
+        {
+            return global::Grpc.Core.ServerServiceDefinition.CreateBuilder()
+                    #{newlineSep 5 serviceMethodBuilder serviceMethods}
+                    .Build();
+        }
+    }
+|]
+      where
+        uniqImplName name = uniqueName (name ++ "_impl") methodNames
+          where
+            methodNames = map methodName serviceMethods
```

Comment:
`methodNames` doesn't need anything from `uniqImpName`. Consider lifting it to be a peer of `uniqImplName`.